all-sdk/autotuning.md

```diff
@@ -93,77 +93,70 @@ trait SlotSupplier {
   /// return a boolean indicating whether a slot was actually obtained or not. In Rust, the future can simply
   /// be dropped if the reservation is no longer desired.
-  async fn reserve_slot(&self);
+  async fn reserve_slot(&self, ctx: &dyn SlotReservationContext<Self::SlotKind>);
 
   /// Tries to immediately reserve a slot, returning true if a slot is available.
-  fn try_reserve_slot(&self) -> bool;
+  fn try_reserve_slot(&self, ctx: &dyn SlotReservationContext<Self::SlotKind>) -> bool;
 
   /// Marks a slot as actually now being used. This is separate from reserving one because the pollers need to
   /// reserve a slot before they have actually obtained work from server. Once that task is obtained (and validated)
   /// then the slot can actually be used to work on the task.
   /// 
   /// Users' implementation of this can choose to emit metrics, or otherwise leverage the information provided by the
   /// `info` parameter to be better able to make future decisions about whether a slot should be handed out.
-  /// 
-  /// `info` may not be provided if the slot was never used
-  /// `error` may be provided if an error was encountered at any point during processing
-  ///     TODO: Error type should maybe also be generic and bound to slot type
-  fn mark_slot_used(&self, info: Option<&Self::SlotKind::Info>, error: Option<&anyhow::Error>);
+  fn mark_slot_used(&self, info: <Self::SlotKind as SlotKind>::Info<'_>);
 
-  /// Frees a slot.
-  fn release_slot(&self, info: &SlotReleaseReason<Self::SlotKind::Info>);
-  
-  /// If this implementation knows how many slots are available at any moment, it should return that here.
-  fn available_slots(&self) -> Option<usize>;
+  /// Frees a slot. This is always called when a slot is no longer needed, even if it was never marked as used.
+  /// EX: If the poller reserves a slot, and then receives an invalid task from the server for whatever reason, this
+  /// method would be called with [SlotReleaseReason::Error].
+  fn release_slot(&self, info: SlotReleaseReason);
 }
 
-struct WorkflowSlotsInfo {
-  used_slots: Vec<WorkflowSlotInfo>,
-  /// Current size of the workflow cache.
-  num_cached_workflows: usize,
-  /// The limit on the size of the cache, if any. This is important for users to know as discussed below in the section
-  /// on workflow cache management.
-  max_cache_size: Option<usize>,
-  // ... Possibly also metric information
-}
-struct ActivitySlotsInfo {
-  used_slots: Vec<ActivitySlotInfo>,
+/// This trait lets implementors obtain other information that might be relevant to their decision on whether to hand
+/// out a slot. It's a trait rather than a struct because the most up-to-date information should be available even
+/// if waiting for some time in the blocking version of `reserve_slot`.
+pub trait SlotReservationContext<SK: SlotKind>: Send + Sync {
+  /// Returns information about currently in-use slots
+  fn used_slots(&self) -> &[SK::Info];
```

Comment:
Shouldn't the slot supplier know this?

Reply:
It *can* know it, but, I don't really see a reason to force the implementor to track it themselves when we already have to track it ourselves at least in some capacity as discussed here: https://github.com/temporalio/proposals/pull/88#discussion_r1502691203

Certainly in Core this made dramatically more sense to provide automatically because the `SlotSupplier` gets wrapped by a struct in Core that has to track all this stuff anyway, and I anticipate creating a similar wrapper in Go/Java.

If it turns out to be a huge pain to provide as I'm working through Java, it can be left off - but I imagine it's going to be pretty straightforward to do.

Reply:
I think just because we can provided something doesn't mean we should. We can always add information based on user feadback, it is much harder to remove. My main concern is the adding the overhead of keeping track of all this data getting updated from potential 10,000s of goroutines in Go when a user doesn't care about the information.

Reply:
Same, I want the lightest API surface needed to do an an acceptable job (may be only slot kind and task queue name atm)

Reply:
There's no question that the information may be valuable, it's who is responsible for it. But I don't have a strong opinion here.

Reply:
If users need to know information about what slots their slot provider handed out they should record that themselves since if they need to record any custom information they will have to handle that themselves anyway

Reply:
Well, sure, but this isn't custom information - it's information that's extracted from the task which we'll be storing anyway. It's just making it available to them rather than asking them to do that work themselves. I don't think there's any strong reason they should record it versus us. It's just making their lives easier for something that's very likely to come up in many or most implementations.

Reply:
The resource based one does need it. It needs to know how many slots are used since you need to know that you're not going to hand out more slots than there is room in the cache (or even just to define a configurable hard limit).

This is zero cost in Core, for example. The info is all tracked already. Fetching system resources like memory usage for that implementation is dramatically more expensive for example.

We can benchmark... but I feel that's a little odd here given there are virtually zero changes we do actually benchmark and I don't see any reason to believe this would add substantial overhead compared to anything else we do. I mean it's literally just stuffing some things (which are likely just references to objects that already exist elsewhere) in a collection, which isn't very expensive to do.

Reply:
>The resource based one does need it. It needs to know how many slots are used since you need to know that you're not going to hand out more slots than there is room in the cache (or even just to define a configurable hard limit).

The resource based one just needs to know the number it doesn't need a whole map of every slot and details about it.

> but I feel that's a little odd here given there are virtually zero changes we do actually benchmark

I get for Core it is zero cost, but for Go and Java this is new. I thought in slack we already talked about benchmarking these changes so I didn't even think this was a new requirement? I assumed as part of benchmarking we would test the old SDK implementation with the new fixed slot supplier implementation?

Reply:
Yeah, of course, but isolating providing this particular function will require it's own independent round of testing - which is fine, but further than we have gone before for any micro-benchmark kind of thing.

>The resource based one just needs to know the number it doesn't need a whole map of every slot and details about it.
Yep - but this is more generally useful and (I would bet quite readily) not meaningfully more costly.

In any case, none of this is really a hard blocker either way on this proposal. We need the context object and it'll have basic stuff like `get_task_queue`. Whether or not this specific function lives on it we can decide after seeing how much effort and overhead it costs, though I'm pretty sure those are both small.